### PR TITLE
Feature/abi merge master again

### PR DESCRIPTION
**Pull request by tzaffi**
This merges master back into `feature/abi`. The only major conflicts were in the in `subroutine.py::SubroutineDefinition`. Validation was conducted in two different ways:
* `feature/abi` - used `_arg_types_and_by_refs`. This has now been removed in favor of:
* `master` - `_validate()`

Admittedly, this is not a fun PR to review in great detail, but something like this was inevitable given all the changes happening on the main branch.

### CODE REVIEW

File: .flake8 [hunk 1]
```diff
@@ -23,9 +24,7 @@ per-file-ignores =
...
     pyteal/__init__.py:                             F401, F403
     pyteal/ir/ops.py:                               E221
-    tests/module_test.py:                           F401, F403
-    tests/*.py:                                     I252
+    tests/unit/module_test.py:                      F401, F403
 
 # from flake8-tidy-imports
-ban-relative-imports = true
-
+ban-relative-imports = true
```
Comment (tzaffi) [inline .flake8:30]:
```suggestion
ban-relative-imports = true

```

File: .github/workflows/build.yml [hunk 2]
```diff
@@ -61,4 +110,4 @@ jobs:
         uses: pypa/gh-action-pypi-publish@release/v1
         with:
           user: __token__
-          password: ${{ secrets.PYPI_API_TOKEN }}
+          password: ${{ secrets.PYPI_API_TOKEN }}
```
Comment (tzaffi) [inline .github/workflows/build.yml:113]:
```suggestion
          password: ${{ secrets.PYPI_API_TOKEN }}

```

File: Makefile [hunk 3]
```diff
@@ -20,30 +27,63 @@ bundle-docs: bundle-docs-clean
...
+
+local-gh-simulate:
+	act
+
+
+# ---- Extras ---- #
+
 coverage:
-	pytest --cov-report html --cov=pyteal
+	pytest --cov-report html --cov=pyteal
```
Comment (tzaffi) [inline Makefile:89]:
```suggestion
	pytest --cov-report html --cov=pyteal

```

File: pyteal/ast/subroutine.py [hunk 4]
```diff
@@ -1,5 +1,5 @@
 from collections import OrderedDict
```
Comment (tzaffi) [inline pyteal/ast/subroutine.py:1]: TODO: Lots'o'merge changes in this file.

**Review by michaeldiamant — APPROVED** [review]
@tzaffi Thanks for working on the merge.

I most closely inspected `subroutine.py` and it looked _reasonable_.  Unless someone spots an obvious error, I'd favor merging and using the branch as a way to further litmus test the merge.

**Review by michaeldiamant — APPROVED** [review]

File: pyteal/ast/subroutine.py [hunk 5]
```diff
@@ -123,7 +123,7 @@ def _validate(
                 f"does not match detected number of parameters ({len(implementation_params)})"
             )
 
-        if "return" in annotations and annotations["return"] is not Expr:
+        if "return" in anns and anns["return"] is not Expr:
```
Comment (ahangsu) [inline pyteal/ast/subroutine.py:126]: what is this...?

**Review by jasonpaulos — APPROVED** [review]
Just a nit, otherwise I have no complaints

**Review by ahangsu — APPROVED** [review]
good to go with Jason's comments resolved

File: pyteal/ast/__init__.py [hunk 6]
```diff
@@ -139,6 +139,7 @@
 
 # abi
 import pyteal.ast.abi as abi  # noqa: I250
+from pyteal.ast.opup import OpUp, OpUpMode
```
[the remark this answers is not in the document]
Comment (ahangsu) [reply pyteal/ast/__init__.py]: I think in the latest commit `opup` is one line above `abi` comment.

**Pull request merged**